### PR TITLE
[RFC FS-1001 POC] String Interpolation [WIP]

### PR DESCRIPTION
**Pull request by yatli**
Planning
---------

This is the beginning of my take on FS-1001 (https://github.com/fsharp/fslang-design/issues/6, https://github.com/fsharp/fslang-design/issues/368) -- I have just started to code it, but I'd like to know what you think.

In this thread, I will interchangeably call the original formatting specifiers "deferred specifier", and inline interpolations "immediate specifier".

### New Tokens and Types

We'll have to introduce new tokens & non-terminals in the parser to accommodate the interpolated string information. Given a string:

image

We want to tokenize the string into:

image

When we parse the token stream:

```
SIP_LEFT SynExpr SIP_MID SynExpr SIP_RIGHT
```

In the parser, we can deduce with the following rules:

```
STRING_INTEROP_HEAD:
| SIP_LEFT SynExpr
| STRING_INTEROP_HEAD SIP_MID SynExpr

STRING_INTEROP:
| STRING_INTEROP_HEAD SIP_RIGHT
```

### Tokenizer Challenges

If we let the hierarchy of nested interpolation slip from the lexer to the parser, we will have to construct ad-hoc parser and start over again. This is not desirable. However, to get a clean token stream considering all the syntax/semantics is a bit challenging. Here's how it's done in Roslyn 
https://github.com/dotnet/roslyn/blob/14e0068d9d529ade86c0ed1544477a221809d0e0/src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs#L222

```C#
            // We have a string of the form
            //                $" ... "
            // or, if isVerbatim is true, of possible forms
            //                $@" ... "
            //                @$" ... "
            // Where the contents contains zero or more sequences
            //                { STUFF }
            // where these curly braces delimit STUFF in expression "holes".
            // In order to properly find the closing quote of the whole string,
            // we need to locate the closing brace of each hole, as strings
            // may appear in expressions in the holes. So we
            // need to match up any braces that appear between them.
            // But in order to do that, we also need to match up any
            // /**/ comments, ' characters quotes, () parens
            // [] brackets, and "" strings, including interpolated holes in the latter.
```
So basically they hoist a non-regular, mini parser that counts braces.

There are two approaches for us:
1. Follow Roslyn
2. Embed interpolation states into the main lexer, so that when an unmatched `}` is encountered, either yield `RBRACE`, or resume the string lexer to yield either `SIP_MID` or `SIP_RIGHT`.

Option 2 is perhaps more maintainable, because it does not drift away from the "real lexer".

### Implementing FS-1001

Suppose we have the token stream properly parsed, and we have `STRING_INTEROP` nodes in the untyped AST. The string formatters are generated at a much later stage, in src/fsharp/CheckFormatStrings.fs, which is only called by the type checker https://github.com/dotnet/fsharp/blob/8397008b8d48f5c2a77fa1ae5fb66d293e8b607c/src/fsharp/TypeChecker.fs#L6851

There are two situations here.
1. ~~The interpolated string does not have a "deferred" specifier.~~
2. The interpolated string has one or more "deferred" specifier.

##### Case 1
~~No deferred specifier means that we don't have to worry about string formatters. Simply move away from the path stringOrKeywordString -> rawConstant -> constant -...-> later picked up by the type checker, and translate our node as appropriate.~~

~~After translation, we should get something like:~~

image

edit: for the completeness of the feature we have to go for case 2.

##### Case 2
The harder case. We will have to make a closure to capture the interpolated expressions, and return a string formatter for the "deferred" parts. 
```F#
let ($) (x: Printf.StringFormat<'a>) = x
let bar = "baz"
// the format string should be StringFormat<string -> string>
let fmt = $ "foo %{bar} %s"
```

https://github.com/dotnet/fsharp/blob/8397008b8d48f5c2a77fa1ae5fb66d293e8b607c/src/fsharp/FSharp.Core/printf.fs#L1519

The runtime support is here.
If we capture the immediate values in the format, we can translate the format string to indicate the index of the captured values -- for example, `%{s:1}` means to extract from `format.Captures.[1]`

Currently, in the front end (printf), I'm putting the captures in the `Format` objects, and I assume it will be populated by the backend-generated code.
`FormatSpecifier` now optionally holds the index of the capture target.
The capture arrays are then forwarded to the ctor of PrintfEnv's.
The builder now handles "Begin-with-capture" arguments on the builder stack. (wip)

The `Printf.Specialization` class is updated with captured value builders. To prevent combinatorial explosion, we restrict that a capture must be the first argument in the argument pack. 
In the main builder, each time an immediate spec is encountered, the pack should be "flushed".
New builders in the families `FinalCaptureX`, `FinalFastCaptureX`, ... , `ChainedCaptureX` , ... , `LittleAChainedCaptureX`, ... are added.
Note that a format specifier can be both immediate capture and special, so we have to extend the main builder for the special form to consider captures.

Your input is much appreciated!

### CODE REVIEW

**Comment by yatli** [comment]
https://github.com/dotnet/fsharp/blob/8397008b8d48f5c2a77fa1ae5fb66d293e8b607c/src/fsharp/pars.fsy#L2641

Looks like this is the master entry for "r-values"?

**Comment by forki** [comment]
@cartermp looks @yatli is Microsoft employee. Can you please resolve this?

**Comment by yatli** [comment]
Anything I should do at the moment?

**Comment by cartermp** [comment]
@yatli you'll need to sign the CLA to allow any PR to be accepted to the repo. @dsyme can speak more towards this approach.

**Comment by yatli** [comment]
I've signed the CLA. Working on `printf.fs`.

File: src/fsharp/xlf/FSComp.txt.es.xlf [hunk 1]
```diff
@@ -109,7 +109,7 @@
       </trans-unit>
       <trans-unit id="followingPatternMatchClauseHasWrongType">
         <source>All branches of a pattern match expression must return values of the same type as the first branch, which here is '{0}'. This branch returns a value of type '{1}'.</source>
-        <target state="translated">Todas las ramas de una expresión de coincidencia de patrón deben devolver valores del mismo tipo. La primera rama devolvió un valor de tipo "{0}", pero esta rama devolvió un valor de tipo "\{1 \}".</target>
+        <target state="new">All branches of a pattern match expression must return values of the same type as the first branch, which here is '{0}'. This branch returns a value of type '{1}'.</target>
```
Comment (yatli) [inline src/fsharp/xlf/FSComp.txt.es.xlf]: 🤔?

**Comment by yatli** [comment]
In `printf.fs, L1271`, why isn't `ContinuationOnStack` maintained accordingly?

File: src/fsharp/xlf/FSComp.txt.es.xlf (hunk 1 above)
Comment (cartermp) [reply src/fsharp/xlf/FSComp.txt.es.xlf]: These get machine-translated into target languages, so no need to worry about this being modified.

Comment (yatli) [reply src/fsharp/xlf/FSComp.txt.es.xlf]: Thanks @cartermp . I'm just wondering why it turns back to English.

File: src/fsharp/FSharp.Core/printf.fs [hunk 2]
```diff
@@ -135,6 +148,36 @@ module internal PrintfImpl =
         
         let parseTypeChar (s: string) i = 
             s.[i], (i + 1)
+
+        let rec parseFormatSpecifier (s:string) (i:int) =
+
+            System.Diagnostics.Debug.Assert(s.[i] = '%', "s.[i] = '%'")
```
Comment (yatli) [inline src/fsharp/FSharp.Core/printf.fs]: when called recursively it should be '{'

File: src/fsharp/xlf/FSComp.txt.es.xlf (hunk 1 above)
Comment (cartermp) [reply src/fsharp/xlf/FSComp.txt.es.xlf]: Just a part of the build.

**Comment by realvictorprm** [comment]
cool stuff :)

**Comment by cartermp** [comment]
Tagging @dsyme for design review

**Comment by yatli** [comment]
@realvictorprm @forki  I'm wondering if we could team up on this?
Considering that we all tried to attack 1001 before 😀

**Comment by realvictorprm** [comment]
@yatli convinced, are you on the F# Slack?

**Comment by yatli** [comment]
@realvictorprm yeah, I've just joined ~~FSF~~ FSSF :D

**Comment by forki** [comment]
Well, that was long time ago ;-)
Any concrete questions or things you want me to do?

Yatao Li <notifications@github.com> schrieb am Do., 20. Juni 2019, 13:18:

**Comment by yatli** [comment]
@forki no worry, one question for now:
Reading through your PR, I see that you've added some tests.
I'm fairly new to the codebase, so I'm wondering if you could give me some pointers -- where to put unit tests for `FSharp.Core` (now it's relevant in the interpolation), where to put end-to-end test files, where compiler tests, etc.

**Comment by dsyme** [comment]
OK, so after years being stalled, it's time take another look at string interpolation from a design perspective. 

The basic conflict is primarily between 
1. a feature `printf "abc %{x} def` that emphasizes "printf"-style formatting v, with strong type and specifier checking
2. a feature `$"abc {x} def"` that emphasizes "C#/.NET-style" formatting (including any specifiers), with weaker type checking for the well-formedness of any specifiers.

When I first came to this area in 2015/16 I thought the feature would basically be modifying printf to allow embedded expressions, i.e (1).  Since then, I've actually used C# string interpolation and my expectations have changed more to (2).  While it makes me feel slightly queasy to have _both_ C#/NET and printf-style format specifiers baked into F# as a language, it may be that **fully** aligning this feature with C# is the right way to go.  

Either way I'm happy to go with `$"..."` as the basic syntax.  It's just so convenient.  

The C# design is here: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated

@cartermp What's your gut feeling about where we should land w.r.t. expressions and specifiers.  The choices are

1. Allow the C# specification `$"The speed of light is {speedOfLight:N3} km/s."`
2. Use printf-style formatting only `$"The speed of light is %0.3f{speedOfLight} km/s."`
3. Allow both
4. Allow (1) but reserve (2), i.e. embedded %..{...} in F# interpolated strings would give a "this is reserved" warning
5. Allow (2) but reserve (1) 

Orthogonally we would have to decide about an implicit conversion to `FormattableString` and `IFormattable`.  Choices are:

1. Allow this by saying "if the known type prior to checking the string literal is FormattableString or IFormattable then the conversion is inserted by the compiler"
2. Don't allow this
3. Reserve this (note, I don't think anything specific is needed to reserve this, since the rule would be based on the known type)

As a strawman I propose (4) and (3) for these two design points.

**Comment by dsyme** [comment]
Note that one major concern is code that uses `sprintf` heavily, with say 20 placeholders.  People will naturally want to move this kind of code to use string interpolation.

However, it's natural that they wont want to lose any type safety.

My concern with supporting only C#-style specifiers is things like `$"abc {x:N3} def"` where `x` is not even numeric, e.g. it's a function missing an argument or something.  In contrast  `$"abc %0.3f{x} def"` would be typesafe.  And if you start with code like

     sprintf "Col1 %0.3f Col2 %0.3f Col3 %0.3f Col4 %08x Col5 %0.3f" (f1 x1) x2 (f3 x3) (f4 x4) x5

then it's easier to progress it to 

     $"Col1 %0.3f{f1 x1} Col2 %0.3f{x2} Col1 %0.3f{f3 x3} Col4 %08x{f4 x4} Col5 %0.3f{x5}"

than to simultaneously convert to less type-safe code using .NET style format specifiers: 

     $"Col1 {f1 x1:N3} Col2 {x2:N3} Col1 {f3:N3} Col4 {f4 x4:Whatever} Col5 {x5:N3}"

**Comment by realvictorprm** [comment]
I prefer keeping the F# formatting specifier for the string interpolation too. Less type safety isn't what should be.

**Comment by baronfel** [comment]
From a user-perspective it would feel odd to me to have to relearn c# formatting syntax for this feature, unless there was a clear benefit. Are there things that c#-style formats can do that printf-style can't?

**Comment by yatli** [comment]
@dsyme my current implementation is as follow:

- String interpolation works without the `$` prefix
- In an interpolated expression `%{interpolated_exp}`, there are two forms:
  1. `%{typedExprWithStaticOptimizationsBlock}` -- translates to a format string with `%A`
  2. `%{format_spec:typedExprWithStaticOptimizationsBlock}` -- translates to a type-checked format string. Here I'm not going with %format_spec{expr} because it's a breaking change.
  3. `$` is an alias for `sprintf`.
  4. The compiler inserts captured value evaluation code before returning the `Format<_,_,_,_,_>`. The `Format<_>` is extended to be able to store the captured values, and thus the captures "disappear" from the final format. Then the format retains the signature as if no values are captured, and therefore, both styles can be mixed.

**Comment by slang25** [comment]
Am I understanding you right.

```fsharp
$"1 + 2 = %{i:3}"
```
becomes:
```fsharp
sprintf "1 + 2 = %i" 3
```
Or have I got that wrong?

**Comment by yatli** [comment]
@slang25 there's an implementation detail here, that `%{i:3}` here will be translated to `%{i:__INDEX__0}`, and `3` is packed into the format string via an intrinsic. Something like:

```
sprintf <| Capture("1 + 2 = %{i:__INDEX__0", 3)
```

**Comment by yatli** [comment]
Turns out,

%{expr:spec} is also breaking change — spec collides with type annotation

**Comment by dsyme** [comment]
> From a user-perspective it would feel odd to me to have to relearn c# formatting syntax for this feature. 

I agree

> $ is an alias for sprintf.

My instinct is to prefer a tighter spec, where `$` is required and must occur exactly before the `"` (no spaces) and is not an alias but is part of the syntax of an interpolated string.. For example, I've no interest in `($)` being used as a first class value.  Also we may want to make other uses of `$` in the future. 

As an aside, it seems reasonable for `$@"..."` to be used for interpolated verbatim strings, and `$"""..."""` for interpolated multi-line strings.  (I don't think we need worry about `$"..."B` for interpolated byte strings)
  
> `%{expr:spec}` is also breaking change — spec collides with type annotation

I assume these are not breaking changes if the `$` is always required.

> Are there things that c#-style formats can do that printf-style can't?

Yes, .NET-style formats can do several things, e.g. [currency, round-trip, percent and custom formats](https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings).  There is also the non-trivial matter of [culture-sensitive formatting](https://docs.microsoft.com/en-us/dotnet/api/system.string.format?view=netframework-4.8#culture-sensitive-formatting).

> unless there was a clear benefit

There are important two benefits - first, F# is (primarily) a .NET language and .NET formatting is a powerful standard that occurs ubiquitously in .NET code samples and is sometimes used in APIs.

Second, in .NET style formatting you don't need to specify the type format at all, so just 

    $"The speed of light is {speedOfLight} km/s."

instead of

    $"The speed of light is %f{speedOfLight} km/s."

The former is certainly easier to use and I think minimally we must support that, and it would be natural to give the identical interpretation to C#, that is identical to `String.Format("{0}", speedOfLight)`.  Fortunately I believe this exactly coincides with `%O{speedOfLight}` with the exception of the formatting of the `null` value which will be the empty string for .NET-style formatting and `<null>` for F#-style formatting.

As part of this, I'm interested to know if there are any cases where the following differ:

    String.Format("{0}", x)
    sprintf "%O" x
    x.ToString()

The only case I know of where these differ is `null`, but if people could double check this I'd be grateful.

```
> String.Format("{0}", (null:obj));;
val it : string = ""

> sprintf "%O" null;;
val it : string = "<null>"

> (null:obj).ToString();;
System.NullReferenceException: Object reference not set to an instance of an object.
   at <StartupCode$FSI_0023>.$FSI_0023.main@()
Stopped due to error
```

File: Session.vim [hunk 3]
```diff
@@ -0,0 +1,87 @@
+let SessionLoad = 1
```
Comment (dsyme) [inline Session.vim]: git ignore this file :)

Comment (yatli) [reply Session.vim]: yeah sure. Just a work-in-progress thing when I move between my machines :)

**Comment by yatli** [comment]
Some thoughts on C#-style format specs: we can extend our format spec char %O so that it can additionally take C#-style specifiers, for example, `%O:N3`

**Comment by dsyme** [comment]
@yatli I took a look through the implementation just now. It's looking good - so the token ization comes next, right?  Let me know if you need help with that, we can pair-program it or discuss on slack.

1. How much easier does this all become if we don't allow any deferred specifiers in the format string.  I think I'd be totally happy with that limitation, if just from a code comprehension point of view

2. Does it seem feasible for support .NET-style format specifiers where `{<interpolationExpression>[,<alignment>][:<formatString>]}` without the leading '%' is allowed in an interpolated string and given the interpretation `String.Format("{0[,<alignment>][:<formatString>]}", interpolationExpression)`, and this processing is done in the middle of the `printf` implementation much as we do for `%O` patterns today

3. Please add $@"..." to for interpolated verbatim strings, and $"""...""" for interpolated multi-line strings to the list of things that need doing.

Thanks!

**Comment by dsyme** [comment]
> Some thoughts on C#-style format specs: we can extend our format spec char %O so that it can additionally take C#-style specifiers, for example, %O:N3 It breaks existing behaviors though.

Yes, I was wondering exactly the same thing - it's interesting because it then sort of unifies the two approaches nicely

Would there be some non-breaking way of doing this?  Perhaps `%(N3)O` or something.

**Comment by thinkbeforecoding** [comment]
Yes that would be super neat for datetimes…

**Comment by yatli** [comment]
Thank you Don! Yes, next step is to start working on the lexer and then eventually make progress to the type checker and codegen, to rendezvous with the formatter side.

What I see as an advantage of the F# formatting way is that it does not have to serialize everything into a string, which is an immutable object in .NET. It interfaces very nicely with all kinds of streams, while the C#-style interpolation is only for strings.
But this is also the tricky part -- the F# way relies on type checking to figure out that a string is indeed a format string. My idea of making `$` standalone is motivated by this -- a minimal mark for the type checker.

Nevertheless, even if we go with `$"` bound together we can still retain the stream-oriented features given my current modifications to printf.fs.

> 1. How much easier does this all become if we don't allow any deferred specifiers in the format string.  I think I'd be totally happy with that limitation, if just from a code comprehension point of view

I think there won't be a big difference, given that the mixed-style could share almost all the code. The difference will be mainly in the tokenizer, and the format spec mini syntax.

> 2. Does it seem feasible for support .NET-style format specifiers where `{<interpolationExpression>[,<alignment>][:<formatString>]}` without the leading '%' is allowed in an interpolated string and given the interpretation `String.Format("{0[,<alignment>][:<formatString>]}", interpolationExpression)`, and this processing is done in the middle of the `printf` implementation much as we do for `%O` patterns today

> Would there be some non-breaking way of doing this? Perhaps %(N3)O or something.

Currently, if no F#-style format spec is given, I'm using `%A`, which is friendlier to F# constructs. But switching over to `%O` also has its strengths -- locale, culture, etc. But ultimately, we'd need a F# way of dealing with localization -- in C#, the interpolated string is NOT a string, but is convertible to:
1. A string, in current culture
2. A FormattableString, where culture info can be further specified.
3. An IFormattable, similar to 2.

If we simply type-check `$"...{interpolate_expr}..."` as a string, the benefits of `%O` is reduced.

The `FormattableString` concept overlaps with F# `Format<_,_,_,_,_>` (that it is not yet a string blob object), so I guess we can just take advantage of that?

If we can work out the alternative mini syntax for extending %O I think it's a good choice -- %O is then the main entry point for .NET formatters, and there can be `x-printf` dialects which can be curried to provide cultural information. win-win :)

So, to wrap it up: `$"....{}..."` should be a new form, where the format specs can start directly with `{`, a short cut for `sprintf ".....%{}..."`;
The latter form is retained, and new dialects can be added so that we can do:
```F#
let fmt = csprintf some_culture_info
fmt "....%{}..."
```


> 3. Please add $@"..." to for interpolated verbatim strings, and $"""...""" for interpolated multi-line strings to the list of things that need doing.

On it.

**Comment by dsyme** [comment]
> If we simply type-check $"...{interpolate_expr}..." as a string, the benefits of `%O` is reduced.
> The `FormattableString` concept overlaps with F# `Format<_,_,_,_,_>` (that it is not yet a string blob object), so I guess we can just take advantage of that?

Yes, I think we can use the approach similar to string and Format w.r.t. the known type, i.e.
* If the known type unifies with `string` then it's a string
* Otherwise, if the known type is `FormattableString` or `Format` then interpret the string to that type
* Otherwise if it subsumes `IFormattable` then interpret the string to that type.

> But switching over to %O also has its strengths -- locale, culture, etc

Yes, using `%O` feels right, partly because it precisely matches C# for `abc {def} hij`

How about this for a mini-spec:

1. $"....{}..." is a new form called an "interpolation string", and can contain either:

   * Neutral fills: `{<interpolationExpression>}`
   * Printf style fills: `%printfFormat{<interpolationExpression>}`
   * .NET-style fills: `{<interpolationExpression>[,<dotnetAlignment>][:<dotnetFormatString>]}`

   Neutral and .NET-style fills are actually shortcut for a `%O` printf pattern `.....%(dotnetAlignment,dotnetFormatString)O{interpolationExpression}..."`.  

   An interpolation string is checked either as a `string`, `FormattableString`, `IFormattable` or `PrintfFormat`.

   Unfilled placeholders may only be present if it is being interpreted as `PrintfFormat`

2. Existing string literals interpreted as `PrintfFormat` strings do not include any interpolation fills - the `$` must be added to get this.

3. Existing string literals may not be interpreted as `FormattableString` nor `IFormattable` - the `$` must be added to get this.

**Comment by yatli** [comment]
I just realized that extending %O will open a gate to new typechecking possibilities -- we can parse the .NET format specs and take that into consideration (e.g. `#.###` must be a floating number).

**Comment by yatli** [comment]
> Existing string literals may not be interpreted as FormattableString nor IFormattable - the $ must be added to get this.

I like this one -- a clean separation.

> Existing string literals interpreted as PrintfFormat strings do not include any interpolation fills - the $ must be added to get this.

May I ask what's your concern on this?
Indeed, I think, mixing captures and deferred specs is definitely bad when there're dozens of them. More chances of off-by-one (or more) errors.

**Comment by cartermp** [comment]
There are a lot of design discussions going on here. Can we capture some of this in an RFC? I think I generally agree with the direction and @dsyme suggestions here. But there are some sub-points that would need a separate design discussion. For example, for verbatim interoplated strings, what about both `$@` and `@$` (as per C# here: https://github.com/dotnet/csharplang/issues/1630)?

Additionally we'd have to adjust tooling to account for understanding when we're in an interoplated context (and complete the `}` with brace completion.

**Comment by yatli** [comment]
@cartermp yes I agree. There are details that are not covered in FS-1001, some of these points are already well-aligned, and some needs further convergence.

I'm listing some of the well aligned points here (may have missed something):

- `$"....{}..."` as the string interpolation form
- Support verbatim/multiline interpolation (we still have to fix down the exact syntax)
- ``%O` by-default for interpolated expressions
- Extend `%O` to incorporate C# formatting specifiers, for both interpolated/conventional strings
- Take cultural info into consideration, which further implies supporting `IFormatProvider`

And after some more thoughts I agree with @dsyme about separating the interpolated form and the normal form. The thing that still doesn't come to its fixed-point in my head is, whether the interpolated form should be a `Format<_,_,_,_,_>`, or `FormattableString` (net46+ only)/`IFormattable`.

The former incorporates well with the printf family, while the latter is more of the "embracing the C# way" style. I prefer the former because, it implies that we plant our support for `IFormatProvider` into the `printf` family, which supports both the interpolated form and the conventional form.
On the other hand, if we decide to check interpolated string into `FormattableString`/`IFormattable`, then there will be a separation between the localization and custom formatting, for conventional/interpolated forms.

Some complexities in extending the printf family:
There is overlap between `IFormatProvider` and F# `%a` specifier, if we plan to extend the printf family, it has to be adjusted accordingly so that the custom formatter receives the format provider, in addition to the contextual writer and the payload.

**Comment by dsyme** [comment]
@cartermp I will rewrite FS-1001 today

**Comment by yatli** [comment]
@dsyme shall we split out the proposal to extend `%O` for conventional strings?

**Comment by dsyme** [comment]
> @dsyme shall we split out the proposal to extend %O for conventional strings?

It would be great to get it all in together, as a complete package, but I think it depends on whether you feel it fits into the implementation and testing reasonably.

**Comment by yatli** [comment]
I'd love to do it altogether -- just want to make sure that this is indeed the plan :)

**Comment by dsyme** [comment]
> The thing that still doesn't come to its fixed-point in my head is, whether the interpolated form should be a `Format<_,_,_,_,_>`, or `FormattableString (net46+ only)/IFormattable`.

Could we make it that  both interpolated and non-interpolated strings are considered compatible with all of these, depending on the known type from type checking?

> > Existing string literals interpreted as PrintfFormat strings do not include any interpolation fills - the $ must be added to get this.
>
> May I ask what's your concern on this? Indeed, I think, mixing captures and deferred specs is definitely bad when there're dozens of them. More chances of off-by-one (or more) errors.

I think it would just really surprise people if non-interoplated strings could ever have interpolation fills when used as printf-style format strings. Also not backwards compatible is it, if the format for interpolation is `%printfFormat{<interpolationExpression>}`?  For example `sprintf "%d{x}" 3` today produces `3{x}`.

> If possible, I'd very much want interpolated strings to incorporate with xprintf family for its stronger verbal functions (output to streams, print to the screen without first combining the pieces, etc)

Yes, I agree with this

> Rule (2) would mean your example would still need $ to interpolate...
>
> if fmt must be followed by a non-curried form, then csprintf becomes a shortcut for FormattableString.Format(some_culture_info), right?

My thinking was that 

* interpolated strings would support interpolation fills and be usable as any of `Format<_,_,_,_,_>`, or `FormattableString` or `IFormattable`.

* regular strings would not support interpolation fills and are also usable as any of `Format<_,_,_,_,_>`, or `FormattableString` or `IFormattable`.

So you could have
```
let fmt = csprintf some_culture_info
fmt $"....%d{xyz}..."
fmt "....%d..." xyz
```

> Additionally we'd have to adjust tooling to account for understanding when we're in an interoplated context (and complete the } with brace completion.

yes

> There is overlap between IFormatProvider and F# %a specifier, if we plan to extend the printf family, it has to be adjusted accordingly so that the custom formatter receives the format provider, in addition to the contextual writer and the payload.

I think pretty much no one uses `%a`.  If we do want to address this would it be possible to use `%z` or the like for an extension point also taking IFormatProvider?  Would that need to be optional in case it hasn't been provided, or is there a default one that would be passed in.

BTW you can find me on F# Software Foundation Slack and we can chat there if you like

File: src/fsharp/FSharp.Core/printf.fs [hunk 4]
```diff
@@ -3,12 +3,22 @@
...
-        member x.Value = value
-
+        
+        internal new(value, caps, types) as this = 
+            PrintfFormat<_,_,_,_>(value)
+            then this.Captures <- caps
+                 this.Types <- types
+
+        member __.Value = value
+        member val internal Captures: obj[] = Unchecked.defaultof<obj[]> with get, set
```
[the remark this answers is not in the document]
Comment (yatli) [reply src/fsharp/FSharp.Core/printf.fs]: yes totally. `set` slipped through my fingers

**Comment by dsyme** [comment]
@yatli Have you been able to make more progress here?  Is there anything we can help with?

**Comment by yatli** [comment]
@dsyme thanks for getting back :)
was interrupted by a minibatch of deadlines, so I have to put the codebase back into my cache again

**Comment by yatli** [comment]
it'll be great if we can do some coding together :)

**Comment by dsyme** [comment]
> it'll be great if we can do some coding together :)

Can you msg me on FSSF Slack?  Maybe late August

**Comment by naughtyGitCat** [comment]
Now is 2020, Is there any new progress on string interpolation? It's quite a practical feature😀

**Comment by varon** [comment]
Please can we get this in. @dsyme / @yatli ?

**Comment by dsyme** [comment]
Hi @yatli, I'm thinking of taking another look at this.  Would you like to still work on it with me for the next iteration?

**Comment by dsyme** [comment]
> Hi @yatli, I'm thinking of taking another look at this. Would you like to still work on it with me for the next iteration?

(I'm starting with prototyping the lexer and parser changes)

**Comment by yatli** [comment]
Definitely!

Sent from Outlook Mobile<https://aka.ms/blhgte>

________________________________

**Comment by dsyme** [comment]
@yatli I'm going to close this PR and start a new one coming from a feature branch `feature/string-interp` in this repo.  You will be able to send work as PRs to that feature branch if you like.

**Comment by dsyme** [comment]
Replaced by https://github.com/dotnet/fsharp/pull/8907

**Comment by yatli** [comment]
sounds great!

**Comment by dsyme** [comment]
Continued at https://github.com/dotnet/fsharp/pull/8907

**Pull request closed without merging**